modules/imgproc/src/filter.simd.hpp

```diff
@@ -489,17 +489,14 @@ struct RowVec_8u32f
             {
                 v_float32 f = vx_setall_f32(_kx[k]);
                 const uchar* src = (const uchar*)_src + i + k * cn;
-                v_uint8 vs = vx_load(src);
-                v_int16 vs_l = v_reinterpret_as_s16(v_expand_low(vs));
-                v_int16 vs_h = v_reinterpret_as_s16(v_expand_high(vs));
-                v_float32 fvs_ll = v_cvt_f32(v_expand_low(vs_l));
-                v_float32 fvs_lh = v_cvt_f32(v_expand_high(vs_l));
-                v_float32 fvs_hl = v_cvt_f32(v_expand_low(vs_h));
-                v_float32 fvs_hh = v_cvt_f32(v_expand_high(vs_h));
-                s0 = v_muladd(fvs_ll, f, s0);
-                s1 = v_muladd(fvs_lh, f, s1);
-                s2 = v_muladd(fvs_hl, f, s2);
-                s3 = v_muladd(fvs_hh, f, s3);
+                v_float32 vs_ll = v_cvt_f32(v_reinterpret_as_s32(vx_load_expand_q(src)));
+                v_float32 vs_lh = v_cvt_f32(v_reinterpret_as_s32(vx_load_expand_q(src + v_float32::nlanes)));
+                v_float32 vs_hl = v_cvt_f32(v_reinterpret_as_s32(vx_load_expand_q(src + 2*v_float32::nlanes)));
+                v_float32 vs_hh = v_cvt_f32(v_reinterpret_as_s32(vx_load_expand_q(src + 3*v_float32::nlanes)));
```

Comment:
@Nicholas-Ho-arm  - @jondea  noticed that this change might not be optimal. Have you re-run the benchmark before replacing the single load + 6 expand with 4 loads + 4 expands? I guess it should not matter much, because the 4 sequences (load +  expand) can be issued in parallel, while the previous version needs to do 1 load , than 2 parallel expands, and finally 4 parallel expands.

What we could do maybe, to increase the level of parallelism of the original version, is to load 2 (or maybe 4) consecutive `v_uint8` from `src` to increase the data being processed by the loop?

@jondea / @alalek / @asmorkalov  - any opinion here?

Reply:
@fpetrogalli Thank you for your comment. I ran the same perf tests after and noticed that there was practically no difference in the numbers so performance wise, I think they are equivalent.

<details>
<summary>Perf tests results with `vx_load_expand_q`</summary>

|Name of Test|PR time|Reference time|x-factor|
|---|:-:|:-:|:-:|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(127x61, 16SC1, (0, 1), BORDER_CONSTANT)|0.027|0.027|0.99|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(127x61, 16SC1, (0, 1), BORDER_REPLICATE)|0.027|0.027|1.00|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(127x61, 16SC1, (0, 2), BORDER_CONSTANT)|0.029|0.029|0.99|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(127x61, 16SC1, (0, 2), BORDER_REPLICATE)|0.028|0.028|0.99|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(127x61, 16SC1, (1, 0), BORDER_CONSTANT)|0.028|0.028|0.99|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(127x61, 16SC1, (1, 0), BORDER_REPLICATE)|0.027|0.027|1.00|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(127x61, 16SC1, (1, 1), BORDER_CONSTANT)|0.026|0.026|1.00|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(127x61, 16SC1, (1, 1), BORDER_REPLICATE)|0.025|0.025|1.00|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(127x61, 16SC1, (2, 0), BORDER_CONSTANT)|0.029|0.029|1.00|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(127x61, 16SC1, (2, 0), BORDER_REPLICATE)|0.028|0.028|1.01|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(127x61, 16SC1, (2, 2), BORDER_CONSTANT)|0.029|0.029|0.99|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(127x61, 16SC1, (2, 2), BORDER_REPLICATE)|0.028|0.028|1.00|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(127x61, 32FC1, (0, 1), BORDER_CONSTANT)|0.035|0.044|0.79|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(127x61, 32FC1, (0, 1), BORDER_REPLICATE)|0.034|0.043|0.80|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(127x61, 32FC1, (0, 2), BORDER_CONSTANT)|0.036|0.045|0.80|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(127x61, 32FC1, (0, 2), BORDER_REPLICATE)|0.035|0.044|0.80|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(127x61, 32FC1, (1, 0), BORDER_CONSTANT)|0.037|0.045|0.82|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(127x61, 32FC1, (1, 0), BORDER_REPLICATE)|0.036|0.045|0.80|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(127x61, 32FC1, (1, 1), BORDER_CONSTANT)|0.035|0.044|0.79|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(127x61, 32FC1, (1, 1), BORDER_REPLICATE)|0.034|0.043|0.80|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(127x61, 32FC1, (2, 0), BORDER_CONSTANT)|0.036|0.045|0.80|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(127x61, 32FC1, (2, 0), BORDER_REPLICATE)|0.035|0.044|0.80|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(127x61, 32FC1, (2, 2), BORDER_CONSTANT)|0.036|0.045|0.81|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(127x61, 32FC1, (2, 2), BORDER_REPLICATE)|0.036|0.044|0.80|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(320x240, 16SC1, (0, 1), BORDER_CONSTANT)|0.117|0.116|1.01|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(320x240, 16SC1, (0, 1), BORDER_REPLICATE)|0.116|0.114|1.02|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(320x240, 16SC1, (0, 2), BORDER_CONSTANT)|0.138|0.137|1.01|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(320x240, 16SC1, (0, 2), BORDER_REPLICATE)|0.136|0.133|1.02|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(320x240, 16SC1, (1, 0), BORDER_CONSTANT)|0.126|0.124|1.02|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(320x240, 16SC1, (1, 0), BORDER_REPLICATE)|0.124|0.123|1.01|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(320x240, 16SC1, (1, 1), BORDER_CONSTANT)|0.108|0.106|1.02|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(320x240, 16SC1, (1, 1), BORDER_REPLICATE)|0.107|0.105|1.02|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(320x240, 16SC1, (2, 0), BORDER_CONSTANT)|0.139|0.137|1.01|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(320x240, 16SC1, (2, 0), BORDER_REPLICATE)|0.138|0.134|1.03|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(320x240, 16SC1, (2, 2), BORDER_CONSTANT)|0.140|0.136|1.03|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(320x240, 16SC1, (2, 2), BORDER_REPLICATE)|0.139|0.135|1.04|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(320x240, 32FC1, (0, 1), BORDER_CONSTANT)|0.220|0.322|0.68|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(320x240, 32FC1, (0, 1), BORDER_REPLICATE)|0.219|0.320|0.68|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(320x240, 32FC1, (0, 2), BORDER_CONSTANT)|0.241|0.347|0.69|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(320x240, 32FC1, (0, 2), BORDER_REPLICATE)|0.237|0.346|0.69|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(320x240, 32FC1, (1, 0), BORDER_CONSTANT)|0.239|0.341|0.70|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(320x240, 32FC1, (1, 0), BORDER_REPLICATE)|0.239|0.340|0.70|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(320x240, 32FC1, (1, 1), BORDER_CONSTANT)|0.219|0.326|0.67|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(320x240, 32FC1, (1, 1), BORDER_REPLICATE)|0.219|0.320|0.69|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(320x240, 32FC1, (2, 0), BORDER_CONSTANT)|0.240|0.348|0.69|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(320x240, 32FC1, (2, 0), BORDER_REPLICATE)|0.238|0.345|0.69|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(320x240, 32FC1, (2, 2), BORDER_CONSTANT)|0.246|0.348|0.71|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(320x240, 32FC1, (2, 2), BORDER_REPLICATE)|0.240|0.346|0.69|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(640x480, 16SC1, (0, 1), BORDER_CONSTANT)|0.403|0.410|0.99|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(640x480, 16SC1, (0, 1), BORDER_REPLICATE)|0.416|0.404|1.03|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(640x480, 16SC1, (0, 2), BORDER_CONSTANT)|0.481|0.475|1.01|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(640x480, 16SC1, (0, 2), BORDER_REPLICATE)|0.482|0.473|1.02|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(640x480, 16SC1, (1, 0), BORDER_CONSTANT)|0.446|0.439|1.02|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(640x480, 16SC1, (1, 0), BORDER_REPLICATE)|0.444|0.442|1.00|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(640x480, 16SC1, (1, 1), BORDER_CONSTANT)|0.382|0.365|1.05|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(640x480, 16SC1, (1, 1), BORDER_REPLICATE)|0.365|0.368|0.99|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(640x480, 16SC1, (2, 0), BORDER_CONSTANT)|0.479|0.473|1.01|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(640x480, 16SC1, (2, 0), BORDER_REPLICATE)|0.487|0.470|1.04|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(640x480, 16SC1, (2, 2), BORDER_CONSTANT)|0.481|0.476|1.01|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(640x480, 16SC1, (2, 2), BORDER_REPLICATE)|0.477|0.475|1.00|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(640x480, 32FC1, (0, 1), BORDER_CONSTANT)|0.823|1.237|0.67|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(640x480, 32FC1, (0, 1), BORDER_REPLICATE)|0.828|1.232|0.67|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(640x480, 32FC1, (0, 2), BORDER_CONSTANT)|0.909|1.324|0.69|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(640x480, 32FC1, (0, 2), BORDER_REPLICATE)|0.896|1.324|0.68|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(640x480, 32FC1, (1, 0), BORDER_CONSTANT)|0.899|1.327|0.68|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(640x480, 32FC1, (1, 0), BORDER_REPLICATE)|0.906|1.320|0.69|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(640x480, 32FC1, (1, 1), BORDER_CONSTANT)|0.821|1.236|0.66|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(640x480, 32FC1, (1, 1), BORDER_REPLICATE)|0.817|1.237|0.66|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(640x480, 32FC1, (2, 0), BORDER_CONSTANT)|0.908|1.325|0.69|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(640x480, 32FC1, (2, 0), BORDER_REPLICATE)|0.904|1.326|0.68|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(640x480, 32FC1, (2, 2), BORDER_CONSTANT)|0.907|1.325|0.68|
|sobelFilter::Size_MatType_dx_dy_Border3x3::(640x480, 32FC1, (2, 2), BORDER_REPLICATE)|0.906|1.321|0.69|

</details>

Reply:
@Nicholas-Ho-arm  - thank you for confirming! :)